pkg/controller/job/job_controller.go

```diff
@@ -875,7 +881,13 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
 			}
 			jobCtx.podsWithDelayedDeletionPerIndex = getPodsWithDelayedDeletionPerIndex(logger, jobCtx)
 		}
+		if canApplySuccessPolicy(jobCtx) {
+			if msg, met := matchSuccessPolicy(logger, job.Spec.SuccessPolicy, *job.Spec.Completions, jobCtx.succeededIndexes); met {
+				jobCtx.finishedCondition = newCondition(batch.JobSuccessCriteriaMet, v1.ConditionTrue, batch.JobReasonSuccessPolicy, msg, jm.clock.Now())
+			}
+		}
 	}
+
 	suspendCondChanged := false
 	// Remove active pods if Job failed.
 	if jobCtx.finishedCondition != nil {
```

Comment:
Actually, we're doing it in the trackJobStatusAndRemoveFinalizers.
@atiratree Does it make sense?

Reply:
I do not us setting it again once we do this:

```go
// Can't declare the Job as finished yet, as there might be remaining
// pod finalizers or pods that are not in the informer's cache yet.
jobCtx.finishedCondition = nil
```

Reply:
Since the Pods have a finalizer, it shouldn't be a problem: The next reconciler would find the same finished Pods and evaluate the successPolicy first.

In general, because of finalizers, the job controller makes lots of API calls, so it's important to batch them as much as possible.

Reply:
> I do not see us setting it again once we do this:

Sorry for my misunderstanding. You're correct, but I think that we want to expect the next reconcile the same as the FilureTarget because I think the failure of terminating pods means that the controller failed to handle the condition transition.

@alculquicondor @mimowo @atiratree WDYT?

Reply:
> Since the Pods have a finalizer, it shouldn't be a problem: The next reconciler would find the same finished Pods and evaluate the successPolicy first.
> 
> In general, because of finalizers, the job controller makes lots of API calls, so it's important to batch them as much as possible.

Oh, what a timing :) Aldo has already said what I wanted to say :)

Reply:
sounds ok to me, if we are okay with a tiny delay

Reply:
IIUC, we're ok with the mechanism.